### PR TITLE
BBEPAM-189: Create internal Context "BBContext"

### PR DESCRIPTION
**Pull request by Ariel-BuildBen**

### CODE REVIEW

File: utils/bb_context.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package utils
```
Comment (testinggithubsucks) [inline utils/bb_context.go:1]: move to package "context"

File: utils/bb_context.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"github.com/buildben/core/interfaces"
+	"sync"
+)
+
+type bbContext struct {
+	keys map[string]interface{}
+	mu   sync.RWMutex
```
Comment (testinggithubsucks) [inline utils/bb_context.go:10]: we can use simple Mutex

File: utils/bb_context.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package utils
+
+import (
+	"github.com/buildben/core/interfaces"
+	"sync"
+)
+
+type bbContext struct {
```
Comment (testinggithubsucks) [inline utils/bb_context.go:8]: buildbenContext

File: utils/bb_context.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	keys map[string]interface{}
+	mu   sync.RWMutex
+}
+
+func NewBBContext() interfaces.BBContext {
```
Comment (testinggithubsucks) [inline utils/bb_context.go:13]: NewBuildbenContext

File: utils/bb_context.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	return &bbContext{}
+}
+
+func (c *bbContext) Set(key string, value interface{}) {
+	c.mu.Lock()
+	if c.keys == nil {
```
Comment (testinggithubsucks) [inline utils/bb_context.go:19]: if value by key not equal nil we just rewrite this value

File: interfaces/bb_context.go [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package interfaces
+
+type BBContext interface {
+	Set(key string, value interface{})
+	Get(key string) (value interface{}, exists bool)
```
Comment (testinggithubsucks) [inline interfaces/bb_context.go:5]: please don't use aliases for returning value

File: utils/bb_context.go (hunk 5 above)
Comment (Ariel-BuildBen) [reply utils/bb_context.go:19]: It's not for a specific key, but for creating the map which saving all the keys. (it's a lazy initialisation for the "keys" property)

File: context/bb_context.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func NewBuildbenContext() interfaces.BuildbenContext {
+	return &buildbenContext{}
+}
+
+func (c *buildbenContext) Set(key string, value interface{}) {
+	c.mu.Lock()
+	if c.keys == nil {
+		c.keys = make(map[string]interface{})
```
Comment (testinggithubsucks) [inline context/bb_context.go:20]: I think it is more logical to do it during object initialization.
in NewBuildbenContext function

File: context/bb_context.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	}
+
+	c.keys[key] = value
+	c.mu.Unlock()
+}
+
+func (c *buildbenContext) Get(key string) (value interface{}, exists bool) {
```
Comment (testinggithubsucks) [inline context/bb_context.go:27]: please don't use aliases for returning parameters

File: context/bb_context.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	value, exists = c.keys[key]
+	c.mu.Unlock()
+	return
+}
+
+func (c *buildbenContext) MustGet(key string) interface{} {
+	if value, exists := c.Get(key); exists {
+		return value
+	}
+	panic("Key \"" + key + "\" does not exist")
```
Comment (testinggithubsucks) [inline context/bb_context.go:38]: let's probably not panic yet.
gin catches such panics and recovery them, we do not have such a mechanism yet

File: context/bb_context.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func (c *buildbenContext) GetString(key string) string {
+	if val, ok := c.Get(key); ok && val != nil {
+		str, _ := val.(string)
```
Comment (testinggithubsucks) [inline context/bb_context.go:43]: what if value does not cast to string ?

File: context/bb_context.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+		return str
+	}
+
+	return ""
+}
+
+func (c *buildbenContext) GetStringSlice(key string) []string {
+	if val, ok := c.Get(key); ok && val != nil {
+		strSlice, _ := val.([]string)
```
Comment (testinggithubsucks) [inline context/bb_context.go:52]: what if value does not cast to string array ?

File: context/bb_context.go [hunk 12]
```diff
@@ -11,46 +12,51 @@ type buildbenContext struct {
...
+	logrus.Errorf("Key \"" + key + "\" does not exist")
+	return nil
 }
 
 func (c *buildbenContext) GetString(key string) string {
 	if val, ok := c.Get(key); ok && val != nil {
-		str, _ := val.(string)
-		return str
+		str, ok := val.(string)
+		if ok {
```
Comment (testinggithubsucks) [inline context/bb_context.go:46]: you can do without nested if

File: context/bb_context.go [hunk 13]
```diff
@@ -11,46 +12,51 @@ type buildbenContext struct {
...
 
 	return ""
 }
 
 func (c *buildbenContext) GetStringSlice(key string) []string {
 	if val, ok := c.Get(key); ok && val != nil {
-		strSlice, _ := val.([]string)
-		return strSlice
+		strSlice, ok := val.([]string)
+		if ok {
```
Comment (testinggithubsucks) [inline context/bb_context.go:57]: you can do without nested if

**Pull request merged**